.devcontainer/launch.sh

```diff
@@ -59,20 +93,220 @@ parse_options() {
     done
 }
 
+# shellcheck disable=SC2155
 launch_docker() {
-    DOCKER_IMAGE=$(grep "image" "${path}/devcontainer.json" | sed 's/.*: "\(.*\)",/\1/')
-    echo "Found image: ${DOCKER_IMAGE}"
-    docker pull ${DOCKER_IMAGE}
-    docker run   \
-        -it --rm \
-        --user coder \
-        --workdir /home/coder/cccl \
-        --mount type=bind,src="$(pwd)",dst='/home/coder/cccl' \
-        ${DOCKER_IMAGE} \
-        /bin/bash
+    local -;
+    set -euo pipefail;
+
+    tabs_to_spaces() {
+        sed $'s/\t/ /g'
+    }
+
+    no_empty_lines() {
+        grep -v -e '^$' || [ "$?" == "1" ]
+    }
+
+    trim_leading_whitespace() {
+        sed 's/^[[:space:]]*//'
+    }
+
+    trim_trailing_whitespace() {
+        sed 's/[[:space:]]*$//'
+    }
+
+    remove_leading_char() {
+        cut -d"$1" -f1 --complement
+    }
+
+    remove_trailing_char() {
+        rev | cut -d"$1" -f1 --complement | rev
+    }
+
+    strip_enclosing_chars() {
+        remove_leading_char "$1" \
+      | remove_trailing_char "$2"
+    }
+
+    read_docker_image_metadata() {
+        docker inspect --type image --format '{{json .Config.Labels}}' "$DOCKER_IMAGE" \
+      | sed -r 's/.*"devcontainer.metadata":[ ]*"(.*[]\)])",?.*/\1/g' \
+      | sed -r 's/\\\"/"/g'
+    }
+
+    json_map_key_to_shell_syntax() {
+        sed -r 's/^"(.*)":[ ]*?([^,]*),?$/\1=\2/g'
+    }
+
+    translate_local_envvars_to_shell_syntax() {
+        sed -r 's/\$\{localEnv:([^\:]*):?(.*)\}/${\1:-\2}/g'
+    }
+
+    inline_local_workspace_folder() {
+        sed "s@\${localWorkspaceFolder}@$(pwd)@g"
+    }
+
+    inline_container_workspace_folder() {
+        sed "s@\${containerWorkspaceFolder}@${WORKSPACE_FOLDER:-}@g"
+    }
+
+    inline_local_workspace_folder_basename() {
+        sed "s@\${localWorkspaceFolderBasename}@$(basename "$(pwd)")@g"
+    }
+
+    transform_to_one_line() {
+        tabs_to_spaces \
+      | trim_leading_whitespace \
+      | trim_trailing_whitespace \
+      | inline_local_workspace_folder \
+      | inline_container_workspace_folder \
+      | inline_local_workspace_folder_basename \
+      | translate_local_envvars_to_shell_syntax \
+      | tr -s '\n' '\t'
+    }
+
+    json_string() {
+        transform_to_one_line \
+      | grep -Po "\"$1\":\s*\"(.*)\"" \
+      | sed -r "s/.*\"$1\":[ ]*\"([^\"]*)\",?.*/\1/g" \
+      | no_empty_lines
+    }
+
+    json_map() {
+        transform_to_one_line \
+      | grep -Po "\"$1\":\s*{(.*?)\s*?}[^\"]" \
+      | strip_enclosing_chars '{' '}' \
+      | tr -s '\t' '\n' \
+      | json_map_key_to_shell_syntax \
+      | no_empty_lines
+    }
+
+    json_array() {
+        transform_to_one_line \
+      | grep -Po "\"$1\":\s*\[(.*?)\s*?\][^\"]" \
+      | strip_enclosing_chars '[' ']' \
+      | tr -s '\t' '\n' \
+      | sed -r 's/", "/"\n"/g' \
+      | sed -r 's/^(.*"),$/\1/' \
+      | no_empty_lines
+    }
```

Comment:
@jrhemstad I'd like your thoughts on this. @trxcllnt implemented this set of bash functions to parse JSON without needing `jq` as a dependency for users of the `launch.sh` script.

Do you have any concerns about requiring `jq` here? We could simplify this implementation quite a bit with it.

Reply:
I'm fine with adding modes to the script that have as many dependencies as we want, so long as there's a mode that has no additional dependencies than what we had before for the manual docker mode.

Reply:
Updated to use `python -c` to parse JSON and select the fields: https://github.com/NVIDIA/cccl/blob/63cc99b5ee456c226ab2c593d316b4ba3a75d857/.devcontainer/launch.sh#L109-L179